src/main/java/seedu/address/storage/JsonAdaptedLesson.java

```diff
@@ -0,0 +1,106 @@
+                             @JsonProperty("date") String date, @JsonProperty("location") String location,
+                             @JsonProperty("weightage") String weightage) {
+        this.code = code;
+        this.type = type;
+        this.date = date;
+        this.location = location;
+        this.weightage = weightage;
+    }
+
+    /**
+     * Converts a given {@code Lesson} into this class for Jackson use.
+     */
+    public JsonAdaptedLesson(Lesson source) {
+        code = source.getCode().code;
+        type = source.getTypeStr();
+        date = source.getDate().format(Task.FORMATTER);
+        location = source.getLocation().value;
+        weightage = Double.toString(source.getWeightage());
+    }
+
+    /**
+     * Converts this Jackson-friendly adapted contact object into the model's {@code Lesson} object.
+     *
+     * @throws IllegalValueException if there were any data constraints violated in the adapted contact.
+     */
+    public Lesson toModelType() throws IllegalValueException {
+
+        if (code == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, Code.class.getSimpleName()));
+        }
+        if (!Code.isValidCode(code)) {
+            throw new IllegalValueException(Code.MESSAGE_CONSTRAINTS);
+        }
+        final Code modelCode = new Code(code);
+
+        if (type == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, Type.class.getSimpleName()));
+        }
+        final Type modelType = Enum.valueOf(Type.class, type);
+
+        if (date == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT,
+                LocalDate.class.getSimpleName()));
+        }
+        try {
+            LocalDate.parse(date, Task.FORMATTER);
+        } catch (DateTimeParseException e) {
+            throw new IllegalValueException(Task.DATE_MESSAGE_CONSTRAINTS);
+        }
+        final LocalDate modelTime = LocalDate.parse(date, Task.FORMATTER);
+
+        if (location == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, Email.class.getSimpleName()));
+        }
+        if (!Address.isValidAddress(location)) {
+            throw new IllegalValueException(Address.MESSAGE_CONSTRAINTS);
+        }
+        final Address modelLocation = new Address(location);
+
+        if (weightage == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, Double.class.getSimpleName()));
+        }
+        try {
+            Double.parseDouble(weightage);
+        } catch (NumberFormatException e) {
+            throw new IllegalValueException(Task.WEIGHTAGE_MESSAGE_CONSTRAINTS);
+        }
+        final double modelWeightage = Double.parseDouble(weightage);
+
+        return new Lesson(modelCode, modelType, modelTime, modelLocation, modelWeightage);
+    }
+}
```

Comment:
Just wondering, is there any better way than manually checking 5 fields to see if they are null or not? I think there is a requireNotNull method as well right?

Reply:
Yes but apart from checking not null, we must also check if the fields fit the specific requirements, which differs for each requirement

Reply:
Okay sure